integration-tests/src/tests/client/features/multinode_test_loop_example.rs

```diff
@@ -0,0 +1,350 @@
+use near_client::test_utils::client_actions_test_utils::{
+    forward_client_messages_from_client_to_client_actions,
+    forward_client_messages_from_network_to_client_actions,
+    forward_client_messages_from_shards_manager,
+    forward_client_messages_from_sync_jobs_to_client_actions,
+};
+use near_client::test_utils::sync_jobs_test_utils::forward_sync_jobs_messages_from_client_to_sync_jobs_actions;
+use near_client::{Client, SyncAdapter, SyncMessage};
+use near_epoch_manager::shard_tracker::{ShardTracker, TrackedConfig};
+use near_epoch_manager::EpochManager;
+use near_network::client::{
+    BlockApproval, BlockResponse, ClientSenderForNetwork, ClientSenderForNetworkMessage,
+};
+use near_network::test_loop::SupportsRoutingLookup;
+use near_network::types::{
+    NetworkRequests, PeerManagerMessageRequest, PeerManagerMessageResponse, SetChainInfo,
+};
+use near_primitives::network::PeerId;
+use near_primitives::shard_layout::ShardLayout;
+use near_primitives::state_record::StateRecord;
+use near_primitives::test_utils::{create_test_signer, create_user_test_signer};
+use near_primitives::types::{AccountId, AccountInfo};
+use near_primitives::version::PROTOCOL_VERSION;
+use near_primitives_core::account::{AccessKey, Account};
+use near_primitives_core::hash::CryptoHash;
+use near_store::genesis::initialize_genesis_state;
+use near_store::test_utils::create_test_store;
+use nearcore::NightshadeRuntime;
+use std::path::Path;
+use std::sync::{Arc, RwLock};
+
+#[derive(derive_more::AsMut, derive_more::AsRef)]
+struct TestData {
+    pub dummy: (),
+    pub account: AccountId,
+    pub client: ClientActions,
+    pub sync_jobs: SyncJobsActions,
+    pub shards_manager: ShardsManager,
+}
+
+impl AsMut<TestData> for TestData {
+    fn as_mut(&mut self) -> &mut Self {
+        self
+    }
+}
+
+#[derive(EnumTryInto, Debug, EnumFrom)]
+#[allow(clippy::large_enum_variant)]
+enum TestEvent {
+    Task(Arc<TestLoopTask>),
+    Adhoc(AdhocEvent<TestData>),
+    ClientDelayedActions(TestLoopDelayedActionEvent<ClientActions>),
+    ClientEventFromNetwork(ClientSenderForNetworkMessage),
+    ClientEventFromClient(ClientSenderForClientMessage),
+    ClientEventFromSyncJobs(ClientSenderForSyncJobsMessage),
+    ClientEventFromShardsManager(ShardsManagerResponse),
+    SyncJobsEventFromClient(SyncJobsSenderForClientMessage),
+    SyncJobsEventFromSyncJobs(SyncJobsSenderForSyncJobsMessage),
+    ShardsManagerRequestFromClient(ShardsManagerRequestFromClient),
+    ClientEventFromStateSyncAdapter(SyncMessage),
+    NetworkMessage(PeerManagerMessageRequest),
+    NetworkMessageForResult(
+        MessageWithCallback<PeerManagerMessageRequest, PeerManagerMessageResponse>,
+    ),
+    SetChainInfo(SetChainInfo),
+}
+
+const ONE_NEAR: u128 = 1_000_000_000_000_000_000_000_000;
+
+// TODO(robin-near): Complete this test so that it will actually run a chain.
+// TODO(robin-near): Make this a multi-node test.
+// TODO(robin-near): Make the network layer send messages.
```

Comment:
Not sure if the second and third todo are relevant any more.
First one remains due to the inability to support rayon thread?

Reply:
Ah yes, I'll delete it next.